### PR TITLE
[mds-db] New database select methods

### PR DESCRIPTION
**Pull request by mdurling**
Borrowing a little inspiration from Entity Framework I created some new database methods for being able to specify shape/type and cardinality when selecting data. I updated a couple of the Audit related queries to illustrate how this might be used. Feedback welcomed.

## Impacts
- [ ] Provider
- [ ] Agency
- [x] Audit
- [ ] Policy
- [ ] Compliance
- [ ] Daily
- [ ] Native
- [ ] Policy Author

### CODE REVIEW

**Comment by janedotx** [comment]
I gotta fly out but I'll take a longer look later. In the meantime, this seems to be an improvement over what we're doing now, but I really think we should move to an ORM in the future and away from building our own SQL queries as much as possible.

**Comment by karcass** [comment]
> we should move to an ORM in the future and away from building our own SQL queries as much as possible.

Agree!  The (ahem) artisanal SQL was a go-as-fast-as-possible shortcut dating back to the original prototype.

**Comment by mdurling** [comment]
> > we should move to an ORM in the future and away from building our own SQL queries as much as possible.
> 
> Agree! The (ahem) artisanal SQL was a go-as-fast-as-possible shortcut dating back to the original prototype.

I don't disagree with either of you. I proposed this as a draft because I felt like this syntax was both more flexible while at the same time being less verbose and therefore perhaps somewhat less error prone. We can consider this an interim step or we can jettison it altogether; either way I found crafting it to be a useful exercise.

File: packages/mds-db/sql-utils.ts [hunk 1]
```diff
@@ -46,6 +65,55 @@ export class MDSPostgresClient extends PostgresClient {
...
+    return result.rows
+  }
+
+  public async selectAll<R extends QueryResultRow>(command: string, values: (string | number)[] = []) {
+    return this.select<R>(command, values)
+  }
+
+  // Runs a query and returns a single selected row as an object of type R. An exception
+  // is thrown if there are no selected rows or more than one selected row.
+  public async selectOne<R extends QueryResultRow>(command: string, values: (string | number)[] = []) {
```
Comment (janedotx) [inline packages/mds-db/sql-utils.ts]: Just curious, I thought you were mostly against throwing an exception if something was not found? Having to remember whether to use `selectOne` or `selectOneOrNull` seems like unnecessary mental overhead.

Comment (mdurling) [reply packages/mds-db/sql-utils.ts]: I'm mostly against not having the choice. There's not much overhead in having the ability to ask for one where you expect there to be exactly one and only one vs when there may be one or none. Anyone familiar with Entity Framework will recognize the semantics.

**Review by janedotx — APPROVED** [review]

**Comment by mdurling** [comment]
Deferring. Looking into ORM integration.

**Pull request closed without merging**